### DISCUSSION TITLE
Thumbnail direct links

### DISCUSSION THREAD

**Discussion by maiatobi** [photoprism/photoprism#3620] [General]
I was about to open my router's port to access my Photoprism instance from outside my house and I noticed that thumbnails are exposed without password, so anybody with the link can access the server. 
I read that this works as expected (https://docs.photoprism.app/developer-guide/media/thumbnails/) and I can understand the reasons this is made in this way but I personally think it is still think is a security problem, I sincerely hope that in the future a better way to retrieve the thumbnails while protected is created. At least that the option is given to the user to choose between faster access or more security (using a session cookie). 

The more clear reason why this should be changed is that in Settings the Adminsitrator has the option to disable sharing, and even if they do that you can still share the thumbnails... granted not the original ones but still the thumbnails are pretty big also..

In the mean time I decided to keep using Photoprism but use a client certificate for each client using traefik. 

Here are some simplified instructions if you are insterested:

1. You need to generate a CA certificate and certificates for your users. This instructions seem to be pretty complete:
`https://jamielinux.com/docs/openssl-certificate-authority/.`

2. After following those steps you will have a ca-chain.pem file and a client_cert.pem and client_key.pem files.

4. Modify your Dynamic Configuration in Traefik to request and verify certificates and set the ca-chain file:
```
  options:
    default:
      clientAuth:
        clientAuthType: RequireAndVerifyClientCert
        # in PEM format. each file can contain multiple CAs.
        caFiles:
          - /certs/ca-chain.pem

```
5. Convert client_cert.pem and client_key.pem into a p12 files that you can add into Firefox (Linux) or in Windows in the Control Panel -> Internet Options -> Content -> Certificates -> Import
```
openssl pkcs12 -export -in client_cert.pem -inkey client_key.pem -out my_client.p12
```
6. If you want the p12 file to be added to Android you must use the legacy flag
```
openssl pkcs12 -export -legacy -in client_cert.pem -inkey client_key.pem -out my_client.android.p12
```

Firefox under Linux will not ask you for the certificate ever again once is added. 
Firefox under Android will not work as it does not yet supports client certificates. The best option in Android seems to be Chrome.
Android Chrome and Windows Edge will ask you to choose the certificate the first time you load the application and access the server.

**Comment by lastzero** [comment]
If you log out, the thumbnail URL won't work anymore. If someone can intercept your traffic, you have a far bigger problem than just the thumbnails. For example, your password will be compromised. With that, you have full access and don't need the thumbnails or a stolen session cookie anymore.

**Reply by lastzero** [reply]
That said, using a client certificate is the right thing to do for enhanced security! 👍

**Reply by maiatobi** [reply]
Hi,
Thanks a lot for your answer. 
I did not test what happens if everybody logs out, if the links do not work anymore that is, but realistically talking there will always be somebody logged into the system so we have to assume that they are always available.
The certificates are nice but are cumbersome, if you need a certificate then sharing is basically for people that are registered users in the server and they already have access to the pictures. For a sporadic share to people outside you will need to send the certificate and instructions which makes it very hard to use.

regards

**Reply by maiatobi** [reply]
Also I have to accept that, perhaps, I am too paranoid. I just tested that the same happens in google photos (although their random link is much, much longer) and although this is not the way I like it for privacy reasons perhaps the performance penalty of using cookies will be too great when the number of pictures grows.

thanks a lot.
